src/bot.ts

```diff
@@ -552,36 +552,63 @@ async function withRetries<T>(
     task: () => Promise<T>,
     signal?: AbortSignal,
 ): Promise<T> {
+    // Set up delays between retries
+    const INITIAL_DELAY = 50; // ms
+    let lastDelay = INITIAL_DELAY;
+
+    // Define error handler
+    /**
+     * Determines the error handling strategy based on various error types.
+     * Sleeps if necessary, and returns whether to retry or rethrow an error.
+     */
+    async function handleError(error: unknown) {
+        let delay = false;
+        let strategy: "retry" | "rethrow" = "rethrow";
+
+        if (error instanceof HttpError) {
+            delay = true;
+            strategy = "retry";
+        } else if (error instanceof GrammyError) {
+            if (error.error_code >= 500) {
+                delay = true;
+                strategy = "retry";
+            } else if (error.error_code === 429) {
+                const retryAfter = error.parameters.retry_after;
+                if (retryAfter !== undefined) {
+                    // sleep directly, don't let this affect the backoff
+                    await sleep(retryAfter, signal);
+                } else {
+                    delay = true;
+                }
+                strategy = "retry";
+            }
+        }
+
+        if (delay) {
+            // Do not sleep for the first retry
+            if (lastDelay !== INITIAL_DELAY) {
+                await sleep(lastDelay, signal);
+            }
+            const oneHour = 1 * 60 * 60 * 1000; // ms
+            lastDelay = Math.min(oneHour, 2 * lastDelay);
```

Comment:
one hour sounds like a lot… yeah its broken, yeah its broken for some time now but an hour is a lot in times of constantly available services. Maybe reduce to 20 min? Poking a server every 20 min is still not that often.

Reply:
I'm using the same interval that Telegram uses when delivering webhooks. But yes, we can reduce it, I don't have too strong opinions here.